File: src/forager/ui/theme.py
from __future__ import annotations
from PySide6.QtGui import QColor, QPalette, QFont
from PySide6.QtWidgets import QApplication

from forager.ui.fonts import UI_FONT


class C:
    """SpaceTheme palette (https://github.com/SpaceTheme/Steam)."""

    BG = "#0a0a0a"
    COLOR_1 = "#111111"
    COLOR_2 = "#1e1e1e"
    COLOR_3 = "#141414"
    COLOR_4 = "#181818"
    COLOR_5 = "#26292c"
    COLOR_6 = "#262629"
    ACCENT_1 = "#666cff"
    ACCENT_2 = "#878cff"
    RED = "#f04a4a"
    RED_HOVER = "#f26363"
    GREEN = "#24a65a"
    GREEN_HOVER = "#27b964"
    BLUE = "#4b89ef"
    BLUE_HOVER = "#649af2"
    YELLOW = "#ef8d4b"
    TEXT = "#ffffff"
    TEXT_DIM = "#8e8e8e"
    TEXT_MUTED = "#a3aab9"

    RADIUS = 8


# Library page background: a very subtle vertical fade from a blue-tinted
# near-black at the top down to pure --st-background, so the page reads as
# layered instead of one flat fill (SpaceTheme keeps the base near-black but
# stacks shelf surfaces on top of it).
PAGE_BG = (
    "qlineargradient(x1:0, y1:0, x2:0, y2:1, "
    "stop:0 #15151b, stop:0.5 #0f0f14, stop:1 #0a0a0a)"
)

# Shelf/panel surface used for the library sections, matching SpaceTheme's
# game shelves: rgb(var(--st-color-2)) with the theme border radius.
PANEL_QSS = f"""
QFrame#Panel {{
    background-color: {C.COLOR_2};
    border-radius: {C.RADIUS}px;
}}
"""


TAB_QSS = f"""
QPushButton {{
    background-color: {C.COLOR_3};
    color: #a9a9a9;
    border: none;
    border-radius: {C.RADIUS}px;
    padding: 6px 16px;
    font-size: 14px;
}}
QPushButton:hover {{
    background-color: {C.COLOR_4};
}}
QPushButton:checked {{
    background-color: {C.COLOR_1};
    color: #ffffff;
}}
"""

NAV_TAB_QSS = f"""
QPushButton {{
    background: transparent;
    border: none;
    border-bottom: 3px solid transparent;
    color: #8b929a;
    font-size: 14px;
    padding: 8px 14px 5px 14px;
}}
QPushButton:hover {{
    color: #ffffff;
}}
QPushButton:checked {{
    color: {C.ACCENT_1};
    border-bottom: 3px solid {C.ACCENT_1};
}}
"""


DISPLAY_SIZES = [
    ("small", "Small", 120, 180),
    ("medium", "Medium", 165, 248),
    ("large", "Large", 250, 375),
]


def resolve_card_size(key: str) -> tuple[int, int]:
    for k, _label, w, h in DISPLAY_SIZES:
        if k == key:
            return (w, h)
    return (165, 248)


# -- Shared QSS fragments used across dialogs and pages -----------------

INPUT_QSS = f"""
QLineEdit {{
    background-color: {C.COLOR_2}; color: {C.TEXT};
    border: 1px solid {C.COLOR_5}; border-radius: {C.RADIUS}px;
    padding: 6px 10px;
}}
QLineEdit:hover {{ border-color: {C.COLOR_6}; }}
QLineEdit:focus {{ border: 1px solid {C.ACCENT_1}; }}
"""

NOTE_QSS = f"color: {C.TEXT_DIM}; font-size: 11px; background: transparent;"

SECTION_TITLE_QSS = f"color: {C.ACCENT_1}; font-size: 13px; font-weight: 800; background: transparent;"

SECTION_QSS = f"""
QFrame#Section {{
    background-color: {C.COLOR_3};
    border: none;
    border-radius: {C.RADIUS}px;
}}
"""

ROW_QSS = f"""
QFrame#Row {{
    background-color: {C.COLOR_2};
    border: none;
    border-radius: 6px;
}}
"""

ROW_LABEL_QSS = f"color: {C.TEXT}; font-size: 13px; font-weight: 500; background: transparent;"

SECONDARY_BTN_QSS = f"""
QPushButton {{ background-color: {C.COLOR_2}; color: {C.TEXT};
 border: 1px solid {C.COLOR_3}; border-radius: {C.RADIUS}px;
 padding: 5px 14px; font-size: 13px; font-weight: 500; }}
QPushButton:hover {{ background-color: {C.COLOR_3}; }}
"""

PRIMARY_BTN_QSS = f"""
QPushButton {{ background-color: {C.ACCENT_1}; color: {C.TEXT}; border: none;
 border-radius: {C.RADIUS}px; padding: 5px 16px; font-size: 13px; font-weight: 600; }}
QPushButton:hover {{ background-color: {C.ACCENT_2}; }}
QPushButton:disabled {{ background-color: {C.COLOR_2}; color: {C.TEXT_DIM}; }}
"""

CHECK_QSS = f"""
QCheckBox {{ color: {C.TEXT}; background: transparent; spacing: 10px; font-size: 13px; }}
QCheckBox::indicator {{
    width: 16px; height: 16px;
    border: 2px solid {C.COLOR_5}; border-radius: 4px;
    background-color: {C.COLOR_3};
}}
QCheckBox::indicator:hover {{ border: 3px solid {C.COLOR_5}; }}
QCheckBox::indicator:checked {{
    background-color: {C.ACCENT_1};
    border: 2px solid {C.ACCENT_1};
    image: none;
}}
"""

RADIO_QSS = f"""
QRadioButton {{
    background-color: {C.COLOR_2}; color: {C.TEXT};
    border: none; border-radius: 6px;
    padding: 8px 12px; font-size: 13px;
    spacing: 10px;
}}
QRadioButton:hover {{ background-color: {C.COLOR_4}; }}
QRadioButton:checked {{
    background-color: {C.ACCENT_1}; color: {C.TEXT};
    font-weight: 600;
}}
QRadioButton::indicator {{
    width: 16px; height: 16px;
    border: 2px solid {C.COLOR_5}; border-radius: 8px;
    background-color: {C.COLOR_3};
}}
QRadioButton::indicator:hover {{ border: 3px solid {C.COLOR_5}; }}
QRadioButton::indicator:checked {{
    border: 2px solid {C.TEXT}; background-color: {C.TEXT};
}}
"""

STATUS_QSS = f"color: {C.TEXT_DIM}; font-size: 11px; background: transparent;"
TITLE_QSS = f"color: {C.TEXT}; font-size: 14px; font-weight: 600; background: transparent;"
LINK_QSS = f"""
QPushButton {{ background: transparent; color: {C.ACCENT_1}; border: none;
 padding: 2px 0; text-align: left; }}
QPushButton:hover {{ color: {C.ACCENT_2}; }}
"""


def load() -> dict[str, str]:
    return {k: getattr(C, k) for k in dir(C) if k.isupper()}


def apply(app: QApplication) -> None:
    app.setStyle("Fusion")

    font = QFont(UI_FONT, 10)
    app.setFont(font)

    p = QPalette()
    bg = QColor(C.BG)
    base = QColor(C.COLOR_1)
    surface = QColor(C.COLOR_2)
    alt = QColor(C.COLOR_3)
    fg = QColor(C.TEXT)
    dim = QColor(C.TEXT_DIM)
    accent = QColor(C.ACCENT_1)

    p.setColor(QPalette.ColorRole.Window, bg)
    p.setColor(QPalette.ColorRole.WindowText, fg)
    p.setColor(QPalette.ColorRole.Base, base)
    p.setColor(QPalette.ColorRole.AlternateBase, alt)
    p.setColor(QPalette.ColorRole.Text, fg)
    p.setColor(QPalette.ColorRole.Button, surface)
    p.setColor(QPalette.ColorRole.ButtonText, fg)
    p.setColor(QPalette.ColorRole.BrightText, accent)
    p.setColor(QPalette.ColorRole.Light, QColor(C.COLOR_5))
    p.setColor(QPalette.ColorRole.Dark, QColor("#000000"))
    p.setColor(QPalette.ColorRole.Mid, alt)
    p.setColor(QPalette.ColorRole.Midlight, QColor(C.COLOR_4))
    p.setColor(QPalette.ColorRole.Shadow, QColor("#000000"))
    p.setColor(QPalette.ColorRole.Highlight, accent)
    p.setColor(QPalette.ColorRole.HighlightedText, QColor(C.BG))
    p.setColor(QPalette.ColorRole.Link, accent)
    p.setColor(QPalette.ColorRole.LinkVisited, QColor(C.ACCENT_2))
    p.setColor(QPalette.ColorRole.PlaceholderText, dim)
    p.setColor(QPalette.ColorRole.ToolTipBase, surface)
    p.setColor(QPalette.ColorRole.ToolTipText, fg)

    app.setPalette(p)
    app.setStyleSheet(stylesheet())


def stylesheet() -> str:
    return f"""
    QMainWindow, QWidget {{
        background-color: {C.BG};
        color: {C.TEXT};
        font-family: "Be Vietnam Pro", "DejaVu Sans", sans-serif;
    }}
    QWidget:disabled {{
        color: {C.TEXT_DIM};
    }}

    QLabel {{
        color: {C.TEXT};
        background: transparent;
    }}

    QPushButton {{
        background-color: {C.COLOR_2};
        color: {C.TEXT};
        border: none;
        border-radius: {C.RADIUS}px;
        padding: 6px 16px;
        font-size: 13px;
        font-weight: 500;
    }}
    QPushButton:hover {{
        background-color: {C.COLOR_3};
    }}
    QPushButton:pressed {{
        background-color: {C.COLOR_4};
    }}
    QPushButton:disabled {{
        background-color: {C.COLOR_2};
        color: {C.TEXT_DIM};
    }}
    QPushButton:focus {{
        outline: none;
    }}

    QLineEdit {{
        background-color: {C.COLOR_3};
        color: {C.TEXT};
        border: none;
        border-radius: {C.RADIUS}px;
        padding: 6px 12px;
        font-size: 13px;
        selection-background-color: {C.ACCENT_1};
        selection-color: {C.BG};
    }}
    QLineEdit:focus {{
        border: 1px solid {C.ACCENT_1};
    }}
    QLineEdit::placeholder {{
        color: {C.TEXT_DIM};
    }}

    QScrollArea {{
        border: none;
        background: transparent;
    }}

    QScrollBar:vertical {{
        background: transparent;
        width: 8px;
        margin: 2px;
    }}
    QScrollBar::handle:vertical {{
        background: {C.COLOR_5};
        border-radius: 4px;
        min-height: 40px;
    }}
    QScrollBar::handle:vertical:hover {{
        background: {C.COLOR_6};
    }}
    QScrollBar::add-line:vertical, QScrollBar::sub-line:vertical {{
        height: 0;
    }}
    QScrollBar:horizontal {{
        background: transparent;
        height: 8px;
        margin: 2px;
    }}
    QScrollBar::handle:horizontal {{
        background: {C.COLOR_5};
        border-radius: 4px;
        min-width: 40px;
    }}
    QScrollBar::handle:horizontal:hover {{
        background: {C.COLOR_6};
    }}
    QScrollBar::add-line:horizontal, QScrollBar::sub-line:horizontal {{
        width: 0;
    }}

    QMenu {{
        background-color: {C.COLOR_2};
        color: {C.TEXT};
        border: none;
        border-radius: {C.RADIUS}px;
        padding: 4px;
        font-size: 13px;
    }}
    QMenu::item {{
        padding: 6px 24px;
        border-radius: 4px;
    }}
    QMenu::item:selected {{
        background-color: {C.COLOR_3};
    }}

    QToolTip {{
        background-color: {C.COLOR_2};
        color: {C.TEXT};
        border: none;
        border-radius: 4px;
        padding: 4px 8px;
    }}

    QMessageBox {{
        background-color: {C.COLOR_1};
    }}
    QMessageBox QLabel {{
        color: {C.TEXT};
    }}
    QMessageBox QPushButton {{
        min-width: 90px;
    }}

    QFileDialog {{
        background-color: {C.COLOR_1};
    }}
    """

File: src/ui/Theme.cpp
#include "ui/Theme.h"
#include "ui/Fonts.h"
#include <QApplication>
#include <QStyleFactory>
#include <QFont>
#include <QPalette>

void theme::applyTheme(QApplication* app) {
    app->setStyle(QStyleFactory::create("Fusion"));

    QFont font(fonts::UI_FONT, 10);
    app->setFont(font);

    QPalette p;
    p.setColor(QPalette::Window, QColor(C::BG));
    p.setColor(QPalette::WindowText, QColor(C::TEXT));
    p.setColor(QPalette::Base, QColor(C::COLOR_1));
    p.setColor(QPalette::AlternateBase, QColor(C::COLOR_3));
    p.setColor(QPalette::Text, QColor(C::TEXT));
    p.setColor(QPalette::Button, QColor(C::COLOR_2));
    p.setColor(QPalette::ButtonText, QColor(C::TEXT));
    p.setColor(QPalette::BrightText, QColor(C::ACCENT_1));
    p.setColor(QPalette::Light, QColor(C::COLOR_5));
    p.setColor(QPalette::Dark, Qt::black);
    p.setColor(QPalette::Mid, QColor(C::COLOR_3));
    p.setColor(QPalette::Midlight, QColor(C::COLOR_4));
    p.setColor(QPalette::Shadow, Qt::black);
    p.setColor(QPalette::Highlight, QColor(C::ACCENT_1));
    p.setColor(QPalette::HighlightedText, QColor(C::BG));
    p.setColor(QPalette::Link, QColor(C::ACCENT_1));
    p.setColor(QPalette::LinkVisited, QColor(C::ACCENT_2));
    p.setColor(QPalette::PlaceholderText, QColor(C::TEXT_DIM));
    p.setColor(QPalette::ToolTipBase, QColor(C::COLOR_2));
    p.setColor(QPalette::ToolTipText, QColor(C::TEXT));

    app->setPalette(p);
    app->setStyleSheet(stylesheet());
}

QString theme::stylesheet() {
    return QStringLiteral(R"(
        QMainWindow, QWidget {
            background-color: %1;
            color: %2;
            font-family: "Be Vietnam Pro", "DejaVu Sans", sans-serif;
        }
        QWidget:disabled { color: %3; }
        QLabel { color: %2; background: transparent; }
        QPushButton {
            background-color: %4; color: %2; border: none;
            border-radius: %6px; padding: 6px 16px; font-size: 13px; font-weight: 500;
        }
        QPushButton:hover { background-color: %5; }
        QPushButton:pressed { background-color: %10; }
        QPushButton:disabled { background-color: %4; color: %3; }
        QPushButton:focus { outline: none; }
        QLineEdit {
            background-color: %5; color: %2; border: none;
            border-radius: %6px; padding: 6px 12px; font-size: 13px;
            selection-background-color: %7; selection-color: %1;
        }
        QLineEdit:focus { border: 1px solid %7; }
        QLineEdit::placeholder { color: %3; }
        QScrollArea { border: none; background: transparent; }
        QScrollBar:vertical { background: transparent; width: 8px; margin: 2px; }
        QScrollBar::handle:vertical { background: %11; border-radius: 4px; min-height: 40px; }
        QScrollBar::handle:vertical:hover { background: %12; }
        QScrollBar::add-line:vertical, QScrollBar::sub-line:vertical { height: 0; }
        QScrollBar:horizontal { background: transparent; height: 8px; margin: 2px; }
        QScrollBar::handle:horizontal { background: %11; border-radius: 4px; min-width: 40px; }
        QScrollBar::handle:horizontal:hover { background: %12; }
        QScrollBar::add-line:horizontal, QScrollBar::sub-line:horizontal { width: 0; }
        QMenu {
            background-color: %4; color: %2; border: none;
            border-radius: %6px; padding: 4px; font-size: 13px;
        }
        QMenu::item { padding: 6px 24px; border-radius: 4px; }
        QMenu::item:selected { background-color: %5; }
        QToolTip {
            background-color: %4; color: %2; border: none;
            border-radius: 4px; padding: 4px 8px;
        }
        QMessageBox { background-color: %13; }
        QMessageBox QLabel { color: %2; }
        QMessageBox QPushButton { min-width: 90px; }
        QFileDialog { background-color: %13; }
    )")
    .arg(C::BG)
    .arg(C::TEXT)
    .arg(C::TEXT_DIM)
    .arg(C::COLOR_2)
    .arg(C::COLOR_3)
    .arg(C::RADIUS)
    .arg(C::ACCENT_1)
    .arg(C::ACCENT_2)
    .arg(C::GREEN)
    .arg(C::COLOR_4)
    .arg(C::COLOR_5)
    .arg(C::COLOR_6)
    .arg(C::COLOR_1);
}
